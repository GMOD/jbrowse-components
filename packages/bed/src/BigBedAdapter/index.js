import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as AdapterClass } from './BigBedAdapter'

export const configSchema = ConfigurationSchema(
  'BigBedAdapter',
  {
    bigBedLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.bb' },
    },
  },
  { explicitlyTyped: true },
)
