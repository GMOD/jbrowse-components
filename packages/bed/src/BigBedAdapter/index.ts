export { default as AdapterClass } from './BigBedAdapter'
export { default as configSchema } from './configSchema'
