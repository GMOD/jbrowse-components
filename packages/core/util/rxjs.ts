import { Observable, Observer } from 'rxjs'
import { takeUntil } from 'rxjs/operators'
import { observeAbortSignal } from '.'

/**
 * Wrapper for rxjs Observable.create with improved error handling and
 * aborting support
 * @param func - observer function, could be async
 */
export function ObservableCreate<T>(
  func: (arg: Observer<T>) => void | Promise<void>,
  signal?: AbortSignal,
): Observable<T> {
  return Observable.create((observer: Observer<T>) => {
    try {
      const ret = func(observer)
      // catch async errors
      if (ret && ret.catch) {
        ret.catch((error: Error) => observer.error(error))
      }
    } catch (error) {
      // catch sync errors
      observer.error(error)
    }
  }).pipe(takeUntil(observeAbortSignal(signal)))
}
