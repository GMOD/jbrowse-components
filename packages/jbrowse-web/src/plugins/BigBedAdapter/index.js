import Plugin from '../../Plugin'
import AdapterType from '../../pluggableElementTypes/AdapterType'
import { ConfigurationSchema } from '../../configuration'
import AdapterClass from './BigBedAdapter'

const configSchema = ConfigurationSchema(
  'BigBedAdapter',
  {
    bigBedLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.bb' },
    },
    assemblyName: {
      type: 'string',
      defaultValue: '',
    },
  },
  { explicitlyTyped: true },
)

export default class BigBedAdapterPlugin extends Plugin {
  install(pluginManager) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'BigBedAdapter',
          configSchema,
          AdapterClass,
        }),
    )
  }
}
