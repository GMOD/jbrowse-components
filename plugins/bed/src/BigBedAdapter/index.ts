import AdapterType from '@jbrowse/core/pluggableElementTypes/AdapterType'

import configSchema from './configSchema'
import type PluginManager from '@jbrowse/core/PluginManager'

export default function BigBedAdapterF(pluginManager: PluginManager) {
  pluginManager.addAdapterType(
    () =>
      new AdapterType({
        name: 'BigBedAdapter',
        displayName: 'BigBed adapter',
        configSchema,
        getAdapterClass: () => import('./BigBedAdapter').then(r => r.default),
      }),
  )
}
