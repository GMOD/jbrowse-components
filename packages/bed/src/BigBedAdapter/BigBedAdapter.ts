/* eslint-disable @typescript-eslint/no-explicit-any */
import { BigBed } from '@gmod/bbi'
import BED from '@gmod/bed'
import {
  BaseFeatureDataAdapter,
  BaseOptions,
} from '@gmod/jbrowse-core/data_adapters/BaseAdapter'
import { Region, FileLocation } from '@gmod/jbrowse-core/util/types'
import { openLocation } from '@gmod/jbrowse-core/util/io'
import { ObservableCreate } from '@gmod/jbrowse-core/util/rxjs'
import SimpleFeature, { Feature } from '@gmod/jbrowse-core/util/simpleFeature'
import { map, mergeAll } from 'rxjs/operators'
import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { Instance } from 'mobx-state-tree'
import configSchema from './configSchema'
import { ucscProcessedTranscript } from '../util'

interface BEDFeature {
  chrom: string
  chromStart: number
  chromEnd: number
  [key: string]: any
}

interface Parser {
  parseLine: (line: string, opts: { uniqueId: string | number }) => BEDFeature
}

export default class BigBedAdapter extends BaseFeatureDataAdapter {
  private bigbed: BigBed

  private parser: Promise<Parser>

  public constructor(config: Instance<typeof configSchema>) {
    super(config)
    const bigBedLocation = readConfObject(
      config,
      'bigBedLocation',
    ) as FileLocation
    this.bigbed = new BigBed({
      filehandle: openLocation(bigBedLocation),
    })

    this.parser = this.bigbed
      .getHeader()
      .then(({ autoSql }: { autoSql: string }) => new BED({ autoSql }))
  }

  public async getRefNames() {
    return Object.keys((await this.bigbed.getHeader()).refsByName)
  }

  public async refIdToName(refId: number) {
    return ((await this.bigbed.getHeader()).refsByNumber[refId] || {}).name
  }

  public getFeatures(region: Region, opts: BaseOptions = {}) {
    const { refName, start, end } = region
    const { signal } = opts
    return ObservableCreate<Feature>(async observer => {
      try {
        const parser = await this.parser
        const ob = await this.bigbed.getFeatureStream(refName, start, end, {
          signal,
          basesPerSpan: end - start,
        })
        ob.pipe(
          mergeAll(),
          map(
            (r: {
              start: number
              end: number
              rest?: string
              uniqueId?: string
            }) => {
              const data = parser.parseLine(
                `${refName}\t${r.start}\t${r.end}\t${r.rest}`,
                {
                  uniqueId: r.uniqueId as string,
                },
              )

              const { blockCount, blockSizes, blockStarts, chromStarts } = data

              if (blockCount) {
                const starts = chromStarts || blockStarts || []
                const sizes = blockSizes
                const blocksOffset = r.start
                data.subfeatures = []

                for (let b = 0; b < blockCount; b += 1) {
                  const bmin = (starts[b] || 0) + blocksOffset
                  const bmax = bmin + (sizes[b] || 0)
                  data.subfeatures.push({
                    uniqueId: `${r.uniqueId}-${b}`,
                    start: bmin,
                    end: bmax,
                    type: 'block',
                  })
                }
              }
              if (r.uniqueId === undefined)
                throw new Error('invalid bbi feature')
              const f = new SimpleFeature({
                id: `${this.id}-${r.uniqueId}`,
                data: {
                  ...data,
                  start: r.start,
                  end: r.end,
                  refName,
                },
              })
              return f.get('thickStart') ? ucscProcessedTranscript(f) : f
            },
          ),
        ).subscribe(observer)
      } catch (e) {
        observer.error(e)
      }
    }, opts.signal)
  }

  public freeResources(): void {}
}
