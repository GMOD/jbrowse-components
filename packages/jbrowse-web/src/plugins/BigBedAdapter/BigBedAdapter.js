import { Observable } from 'rxjs'

import { BigBed } from '@gmod/bbi'
import BED from '@gmod/bed'

import BaseAdapter from '../../BaseAdapter'
import { openLocation } from '../../util/io'
import SimpleFeature from '../../util/simpleFeature'

export default class BigBedAdapter extends BaseAdapter {
  static capabilities = ['getFeatures', 'getRefNames']

  constructor(config) {
    super(config)
    this.bigbed = new BigBed({
      filehandle: openLocation(config.bigBedLocation),
    })

    this.parser = this.bigbed
      .getHeader()
      .then(({ autoSql }) => new BED({ autoSql }))
  }

  async getRefNames() {
    const header = await this.bigbed.getHeader()
    return Object.keys(header.refsByName)
  }

  async refIdToName(refId) {
    return ((await this.bigbed.getHeader()).refsByNumber[refId] || {}).name
  }

  async getGlobalStats() {
    const header = await this.bigbed.getHeader()
    return header.totalSummary
  }

  /**
   * Fetch features for a certain region
   * @param {Region} param
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */
  getFeatures({ /* assembly, */ refName, start, end }) {
    return Observable.create(async observer => {
      const parser = await this.parser
      const records = await this.bigbed.getFeatures(refName, start, end)
      records.forEach(r => {
        const data = parser.parseBedText(refName, r.start, r.end, r.rest, {
          offset: 3,
          uniqueId: r.uniqueId,
        })
        observer.next(new SimpleFeature({ data }))
      })
      observer.complete()
    })
  }
}
