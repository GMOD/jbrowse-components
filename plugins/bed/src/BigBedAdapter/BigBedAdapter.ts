import { BigBed } from '@gmod/bbi'
import BED from '@gmod/bed'
import { BaseFeatureDataAdapter } from '@jbrowse/core/data_adapters/BaseAdapter'
import { doesIntersect2, max, min, SimpleFeature } from '@jbrowse/core/util'
import { openLocation } from '@jbrowse/core/util/io'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import { firstValueFrom, toArray } from 'rxjs'

// locals
import { featureData2 } from '../util'
import type { BaseOptions } from '@jbrowse/core/data_adapters/BaseAdapter'
import type { Feature, SimpleFeatureSerialized } from '@jbrowse/core/util'
import type { Region } from '@jbrowse/core/util/types'
import type { Observer } from 'rxjs'

export default class BigBedAdapter extends BaseFeatureDataAdapter {
  private cachedP?: Promise<{
    bigbed: BigBed
    header: Awaited<ReturnType<BigBed['getHeader']>>
    parser: BED
  }>

  public async configurePre(opts?: BaseOptions) {
    const pm = this.pluginManager
    const bigbed = new BigBed({
      filehandle: openLocation(this.getConf('bigBedLocation'), pm),
    })
    const header = await bigbed.getHeader(opts)
    const parser = new BED({
      autoSql: header.autoSql,
    })
    return {
      bigbed,
      header,
      parser,
    }
  }

  public async configure(opts?: BaseOptions) {
    if (!this.cachedP) {
      this.cachedP = this.configurePre(opts).catch((e: unknown) => {
        this.cachedP = undefined
        throw e
      })
    }
    return this.cachedP
  }

  public async getRefNames(opts?: BaseOptions) {
    const { header } = await this.configure(opts)
    return Object.keys(header.refsByName)
  }

  public async getData() {
    const refNames = await this.getRefNames()
    const features = []
    for (const refName of refNames) {
      const f = await firstValueFrom(
        this.getFeatures({
          assemblyName: 'unknown',
          refName,
          start: 0,
          end: Number.MAX_SAFE_INTEGER,
        }).pipe(toArray()),
      )
      features.push(f)
    }
    return features.flat()
  }

  async getHeader(opts?: BaseOptions) {
    const { parser, header } = await this.configure(opts)
    const { version, fileType } = header
    const { fields, ...rest } = parser.autoSql
    return {
      version,
      fileType,
      autoSql: { ...rest },
      fields: Object.fromEntries(
        fields.map(({ name, comment }) => [name, comment]),
      ),
    }
  }

  public async getFeaturesHelper({
    query,
    opts,
    observer,
    allowRedispatch,
    originalQuery = query,
  }: {
    query: Region
    opts: BaseOptions
    observer: Observer<Feature>
    allowRedispatch: boolean
    originalQuery?: Region
  }) {
    const { stopToken } = opts
    const scoreColumn = this.getConf('scoreColumn')
    const aggregateField = this.getConf('aggregateField')
    const { parser, bigbed } = await this.configure(opts)
    const feats = await bigbed.getFeatures(
      query.refName,
      query.start,
      query.end,
      {
        stopToken,
        basesPerSpan: query.end - query.start,
      },
    )
    if (allowRedispatch && feats.length) {
      let minStart = Number.POSITIVE_INFINITY
      let maxEnd = Number.NEGATIVE_INFINITY
      for (const feat of feats) {
        if (feat.start < minStart) {
          minStart = feat.start
        }
        if (feat.end > maxEnd) {
          maxEnd = feat.end
        }
      }
      if (maxEnd > query.end || minStart < query.start) {
        await this.getFeaturesHelper({
          query: {
            ...query,
            start: minStart,
            end: maxEnd,
          },
          opts,
          observer,
          allowRedispatch: false,
          originalQuery: query,
        })
        return
      }
    }

    const parentAggregation = {} as Record<string, SimpleFeatureSerialized[]>

    if (feats.some(f => f.uniqueId === undefined)) {
      throw new Error('found uniqueId undefined')
    }
    for (const feat of feats) {
      const line = `${query.refName}\t${feat.start}\t${feat.end}\t${feat.rest}`
      const data = parser.parseLine(line, { uniqueId: feat.uniqueId! })

      const aggr = data[aggregateField]
      if (!parentAggregation[aggr]) {
        parentAggregation[aggr] = []
      }
      const {
        uniqueId,
        type,
        chrom,
        chromStart,
        chromEnd,
        description,
        chromStarts: chromStarts2,
        blockStarts: blockStarts2,
        blockSizes: blockSizes2,
        score: score2,
        blockCount,
        thickStart,
        thickEnd,
        strand,
        ...rest
      } = data

      const f = featureData2({
        ...rest,
        scoreColumn,
        line,
        parser,
        uniqueId,
        start: feat.start,
        end: feat.end,
        refName: query.refName,
      })
      if (aggr) {
        parentAggregation[aggr].push(f)
      } else {
        if (
          doesIntersect2(f.start, f.end, originalQuery.start, originalQuery.end)
        ) {
          observer.next(
            new SimpleFeature({
              id: `${this.id}-${uniqueId}`,
              data: f,
            }),
          )
        }
      }
    }

    Object.entries(parentAggregation).map(([name, subfeatures]) => {
      const s = min(subfeatures.map(f => f.start))
      const e = max(subfeatures.map(f => f.end))
      if (doesIntersect2(s, e, originalQuery.start, originalQuery.end)) {
        const { uniqueId, strand } = subfeatures[0]!
        observer.next(
          new SimpleFeature({
            id: `${this.id}-${uniqueId}-parent`,
            data: {
              type: 'gene',
              subfeatures,
              strand,
              name,
              start: s,
              end: e,
              refName: query.refName,
            },
          }),
        )
      }
    })
    observer.complete()
  }
  public getFeatures(query: Region, opts: BaseOptions = {}) {
    return ObservableCreate<Feature>(async observer => {
      try {
        await this.getFeaturesHelper({
          query,
          opts,
          observer,
          allowRedispatch: true,
        })
      } catch (e) {
        observer.error(e)
      }
    }, opts.stopToken)
  }

  public freeResources(): void {}
}
