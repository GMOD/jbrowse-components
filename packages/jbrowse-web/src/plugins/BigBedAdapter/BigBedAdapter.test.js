import { toArray } from 'rxjs/operators'

import BigBedAdapter from './BigBedAdapter'

test('adapter can fetch features from volvox.bb', async () => {
  const adapter = new BigBedAdapter(
    {
      assemblyName: 'volvox',
      bigBedLocation: { localPath: require.resolve('./test_data/volvox.bb') },
    },
    {},
  )

  const features = await adapter.getFeatures({
    assemblyName: 'volvox',
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })
  expect(await adapter.refIdToName(0)).toBe('ctgA')
  expect(await adapter.refIdToName(1)).toBe(undefined)
  expect(
    await adapter.hasDataForRefSeq({ assemblyName: 'volvox', refName: 'ctgA' }),
  ).toBe(true)
  expect(
    await adapter.hasDataForRefSeq({ assemblyName: 'volvox', refName: 'ctgB' }),
  ).toBe(false)
  expect(
    await adapter.hasDataForRefSeq({
      assemblyName: 'nonexist',
      refName: 'ctgA',
    }),
  ).toBe(false)

  const featuresArray = await features.pipe(toArray()).toPromise()
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray.slice(0, 5000)).toMatchSnapshot()
})
