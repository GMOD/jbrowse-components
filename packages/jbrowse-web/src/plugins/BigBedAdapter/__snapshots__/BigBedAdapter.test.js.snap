// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from volvox.bw 1`] = `Array []`;
